module/builder/consensus/builder.go

```diff
@@ -101,18 +101,18 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		return nil, fmt.Errorf("could not insert guarantees: %w", err)
 	}
 
-	// get the seals to insert in the payload
-	insertableSeals, err := b.getInsertableSeals(parentID)
-	if err != nil {
-		return nil, fmt.Errorf("could not insert seals: %w", err)
-	}
-
 	// get the receipts to insert in the payload
 	insertableReceipts, err := b.getInsertableReceipts(parentID)
 	if err != nil {
 		return nil, fmt.Errorf("could not insert receipts: %w", err)
 	}
 
+	// get the seals to insert in the payload
+	insertableSeals, err := b.getInsertableSeals(parentID, insertableReceipts)
+	if err != nil {
+		return nil, fmt.Errorf("could not insert seals: %w", err)
+	}
+
```

Comment:
⚠️ 
Sealing a previously unknown result in the same block as we incorporate the result for the first time is _incompatible_ with the sealing protocol. 

I don't understand how this logic would increase the speed of sealing (?):
* speed of sealing: how many results can we seal per block we construct? 
   * we can always seal the results from the parent
* restricting the sealing to results that have been incorporated in ancestors adds a latency of 1 additional block. However, the sealing speed should remain invariant imo.

Reply:
>we can always seal the results from the parent

Not always, because the leader A puts sealable receipts in the block, and the next leader B might not have the seals for these receipts. It's only guaranteed when the leader A becomes leader again, those seals could be put in the block, but what would be after a while. If only A has the next seal, then in that case sealing won't halt, but get slowed down a lot.

Just to point out this case and the behavior. I still agree we should follow the mature protocol to not seal the result in the same block.

I will revert my change.